middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml

```diff
@@ -495,99 +505,94 @@ let checked_string_or_bytes_access ~dbg ~size_int ~access_size ~primitive kind
   checked_access ~dbg ~primitive
     ~conditions:
       [ string_or_bytes_access_validity_condition ~size_int string kind
-          access_size index ]
+          access_size ~index_kind index ]
 
-let checked_bigstring_access ~dbg ~size_int ~access_size ~primitive arg1 arg2 =
+let checked_bigstring_access ~dbg ~size_int ~access_size ~primitive arg1
+    ~index_kind arg2 =
   let primitive =
     match (access_size : P.string_accessor_width) with
     | One_twenty_eight { aligned = true } ->
       checked_alignment ~dbg ~primitive
-        ~conditions:[bigstring_alignment_validity_condition arg1 16 arg2]
+        ~conditions:
+          [ bigstring_alignment_validity_condition arg1 16
+              (convert_index_to_tagged_int arg2 index_kind) ]
     | Eight | Sixteen | Thirty_two | Single | Sixty_four
     | One_twenty_eight { aligned = false } ->
       primitive
   in
   checked_access ~dbg ~primitive
     ~conditions:
-      [bigstring_access_validity_condition ~size_int arg1 access_size arg2]
+      [ bigstring_access_validity_condition ~size_int arg1 access_size
+          ~index_kind arg2 ]
 
 (* String-like loads *)
-let string_like_load_unsafe ~access_size kind mode ~boxed string index
-    ~current_region =
-  let wrap =
-    match (access_size : Flambda_primitive.string_accessor_width), mode with
-    | (Eight | Sixteen), None ->
-      assert (not boxed);
-      tag_int
-    | Thirty_two, Some mode ->
-      if boxed then box_bint Pint32 mode ~current_region else Fun.id
-    | Single, Some mode ->
-      if boxed then box_float32 mode ~current_region else Fun.id
-    | Sixty_four, Some mode ->
-      if boxed then box_bint Pint64 mode ~current_region else Fun.id
-    | One_twenty_eight _, Some mode ->
-      if boxed then box_vec128 mode ~current_region else Fun.id
-    | (Eight | Sixteen), Some _
-    | (Thirty_two | Single | Sixty_four | One_twenty_eight _), None ->
-      Misc.fatal_error "Inconsistent alloc_mode for string or bytes load"
+let string_like_load ~dbg ~unsafe ~access_size ~size_int kind mode ~boxed string
+    ~index_kind index ~current_region =
+  let unsafe_load =
+    let index = convert_index_to_tagged_int index index_kind in
+    let wrap =
+      match (access_size : Flambda_primitive.string_accessor_width), mode with
```

Comment:
I think in general it's neater to put these annotations on the function parameters, then they never need to be repeated (not that they are here).